### PR TITLE
fix(Item): hide all variants of when template is hidden

### PR DESCRIPTION
**Pull request by hrwX**
-    [Asana Task](https://app.asana.com/0/1188592895233593/1191837972966645)
-    When an Item Template was to be hidden from website, on the uncheck of `show_in_website` triggered `update_item_variants`.
-    When Item Variants are being updated, `show_in_website` property was added to `ignore_fields`, and it triggered `update_template_item` which in turn again checked `show_in_website` for template item.

### CODE REVIEW

**Comment by Alchez** [comment]
@hrwX, which task is this change for? And can we use Item Variant Settings for doing this?

**Comment by hrwX** [comment]
> @hrwX, which task is this change for? And can we use Item Variant Settings for doing this?

-    `show_in_website, show_variant_in_website` are excluded in Item Variant Settings.
-    I have updated the task from asasna

File: erpnext/stock/doctype/item/item.py [hunk 1]
```diff
@@ -489,6 +489,15 @@ def update_show_in_website(self):
...
 			self.show_in_website = False
 
+		variants = frappe.get_all("Item", filters={"variant_of": self.name}, fields=["item_code"])
+
+		if not self.disabled and self.has_variants and not self.show_in_website and variants:
+			frappe.db.sql("""
+				UPDATE `tabItem`
+				SET show_variant_in_website=0
+				WHERE variant_of=%s
+			""", (self.name))
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]: Avoid using SQL updates, that bypasses any other validations that might need to happen (now or later) and sets it directly in the db. That might cause unexpected behaviour.

```suggestion
		if self.disabled:
			self.show_in_website = False
		elif self.has_variants and not self.show_in_website:
			variants = frappe.get_all("Item", filters={"variant_of": self.name})
			for item in variants:
				item_doc = frappe.get_doc("Item", item.name)
				item_doc.show_variant_in_website = False
				item_doc.save()
```

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.py (hunk 1 above)
Comment (hrwX) [reply erpnext/stock/doctype/item/item.py]: So will have to add this to background jobs if variants are more than 30, it'll take forever to save the doc.

Comment (Alchez) [reply erpnext/stock/doctype/item/item.py]: That's fair, I think we can safely move this to background jobs

File: erpnext/stock/doctype/item/item.py [hunk 2]
```diff
@@ -486,9 +486,22 @@ def add_default_uom_in_conversion_factor_table(self):
...
 		if self.disabled:
 			self.show_in_website = False
 
+		elif not self.disabled and self.has_variants and not self.show_in_website and not self.show_in_website == doc.show_in_website:
+			variants = frappe.get_all("Item", filters={"variant_of": self.name}, fields=["item_code"])
+
+			if not variants:
+				return
+
+			if len(variants) <= 30:
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]: Just enqueue it always, there's no visible feedback either way. We don't want to keep adjusting this value if required.

File: erpnext/stock/doctype/item/item.py [hunk 3]
```diff
@@ -486,9 +486,22 @@ def add_default_uom_in_conversion_factor_table(self):
 		[self.remove(d) for d in to_remove]
 
 	def update_show_in_website(self):
+		doc = self.get_doc_before_save()
+
 		if self.disabled:
 			self.show_in_website = False
 
+		elif not self.disabled and self.has_variants and not self.show_in_website and not self.show_in_website == doc.show_in_website:
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]:
```suggestion
		elif self.has_variants and not self.show_in_website and self.show_in_website != doc.show_in_website:
```

File: erpnext/stock/doctype/item/item.py [hunk 4]
```diff
@@ -486,9 +486,22 @@ def add_default_uom_in_conversion_factor_table(self):
...
+			variants = frappe.get_all("Item", filters={"variant_of": self.name}, fields=["item_code"])
+
+			if not variants:
+				return
+
+			if len(variants) <= 30:
+				hide_variants(variants)
+			else:
+				frappe.enqueue("erpnext.stock.doctype.item.item.hide_variants", variants=variants, timeout=600)
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]:
```suggestion
			if variants:
				frappe.enqueue("erpnext.stock.doctype.item.item.hide_variants", variants=variants, timeout=600)
```

**Review by Alchez — CHANGES_REQUESTED** [review]

**Comment by hrwX** [comment]
@Alchez Now that I think, enqueuing hiding of item variants will not fix this issue, since this issue is caused due to the item variants being visible on website in the first place.
At the moment, using a single sql to disable all item variants seems like a better solution as we would not know how many item variants are present.

**Comment by Alchez** [comment]
@hrwX, doing a direct SQL query is always the easiest solution lol, but that tends to cause more problems than it solves by bypassing all controller validations

What's the main problem with enqueueing? If we don't want to enqueue, and have the user wait for all the records to save, maybe we can show a progress bar while running it on the server?

**Comment by hrwX** [comment]
> @hrwX, doing a direct SQL query is always the easiest solution lol, but that tends to cause more problems than it solves by bypassing all controller validations

I know sql is the easiest way out of this issue, but afaik, we are just unchecking the show on website check using sql. Bypassing controllers would be not an ideal solution I get it.  

> What's the main problem with enqueueing? If we don't want to enqueue, and have the user wait for all the records to save, maybe we can show a progress bar while running it on the server?

The problem with enqueuing is that, unless all the variants of the item template are hidden from website, the show on website of item variant cannot be unchecked. So putting the hiding of item variants in enqueue, wont fix this issue. Only way is to hide them is before saving the the item template. My worry is, if there too many variants, it'll take too much time

**Comment by DeveloperMetal** [comment]
@hrwX Currently we don't have a customer with hundreds of variants that I know of. So for the time being this isn't an issue for us but the concern is justified. We do need controllers to handle validations and firing hooks in general.

@Alchez Enqueuing does break the order in which he'll have to do work to disable the items, since the variants will be updated a long while after this item's validation has happened(async). All of this stuff has to happen in order so I do don't think enqueueing is he right answer here either due to the item controller forcing the show_on_website back on the template item.

I only see two options here:

1) Find where in the Item doctype/controller erpn is setting the template item's show_on_website back on when any of its variants are enabled and decouple that. So then you can enqueue your changes and move on.

2) Update items sync starting from variants then finally the template item.

Option 1 seems more scaleable to me if we want to be able to update hundreds of variants eventually.
Option 2 is the dirtiest one, but easy to achieve. Tho as pointed out if we have a lot of variants this will negatively affect user experience.

**Comment by hrwX** [comment]
> @hrwX Currently we don't have a customer with hundreds of variants that I know of. So for the time being this isn't an issue for us but the concern is justified. We do need controllers to handle validations and firing hooks in general.
> 
> @Alchez Enqueuing does break the order in which he'll have to do work to disable the items, since the variants will be updated a long while after this item's validation has happened(async). All of this stuff has to happen in order so I do don't think enqueueing is he right answer here either due to the item controller forcing the show_on_website back on the template item.
> 
> I only see two options here:
> 
>     1. Find where in the Item doctype/controller erpn is setting the template item's show_on_website back on when any of its variants are enabled and decouple that. So then you can enqueue your changes and move on.
> 
>     2. Update items sync starting from variants then finally the template item.
> 
> 
> Option 1 seems more scaleable to me if we want to be able to update hundreds of variants eventually.
> Option 2 is the dirtiest one, but easy to achieve. Tho as pointed out if we have a lot of variants this will negatively affect user experience.

@DeveloperMetal and @Alchez I have decoupled the show_in_variant for item template and item variant so that individual variant can be updated

File: erpnext/stock/doctype/item/item.py [hunk 5]
```diff
@@ -486,9 +485,17 @@ def add_default_uom_in_conversion_factor_table(self):
 		[self.remove(d) for d in to_remove]
 
 	def update_show_in_website(self):
+		show_in_website = frappe.db.get_value("Item", self.name, "show_in_website")
+
 		if self.disabled:
 			self.show_in_website = False
 
+		elif self.has_variants and not self.show_in_website and not self.show_in_website == show_in_website:
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]:
```suggestion
		elif self.has_variants and self.show_in_website != show_in_website:
```

File: erpnext/stock/doctype/item/item.py [hunk 6]
```diff
@@ -486,9 +485,17 @@ def add_default_uom_in_conversion_factor_table(self):
...
+			variants = frappe.get_all("Item", filters={"variant_of": self.name}, fields=["item_code"])
+
+			if variants:
+				frappe.enqueue("erpnext.stock.doctype.item.item.hide_variants", variants=variants, timeout=600)
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py]: Do the same thing for showing the variants too (0 -> 1)

**Review by Alchez — CHANGES_REQUESTED** [review]

File: erpnext/stock/doctype/item/item.py [hunk 7]
```diff
@@ -486,9 +485,17 @@ def add_default_uom_in_conversion_factor_table(self):
 		[self.remove(d) for d in to_remove]
 
 	def update_show_in_website(self):
+		show_in_website = frappe.db.get_value("Item", self.name, "show_in_website")
+
 		if self.disabled:
 			self.show_in_website = False
```
Comment (Alchez) [inline erpnext/stock/doctype/item/item.py:490]: Hide variants if the item is disabled as well

**Review by Alchez — APPROVED** [review]

**Pull request merged**